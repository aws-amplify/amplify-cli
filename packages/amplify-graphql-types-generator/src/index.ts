import './polyfills';

export { default as generate } from './generate';
