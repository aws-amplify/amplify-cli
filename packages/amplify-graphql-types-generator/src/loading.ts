import * as fs from 'fs'

import {
  buildClientSchema,
  Source,
  concatAST,
  parse,
  DocumentNode,
  GraphQLSchema
} from 'graphql';

import { ToolError } from './errors'

export function loadSchema(schemaPath: string): GraphQLSchema {
  if (!fs.existsSync(schemaPath)) {
    throw new ToolError(`Cannot find GraphQL schema file: ${schemaPath}`);
  }
  const schemaData = require(schemaPath);

  if (!schemaData.data && !schemaData.__schema) {
    throw new ToolError('GraphQL schema file should contain a valid GraphQL introspection query result');
  }
  return buildClientSchema((schemaData.data) ? schemaData.data : schemaData);
}

function extractDocumentFromJavascript(content: string, tagName: string = 'gql'): string | null {
  const re = new RegExp(tagName + '\\s*`([^`/]*)`', 'g');

  let match
  const matches = []

  while(match = re.exec(content)) {
    const doc = match[1]
      .replace(/\${[^}]*}/g, '')

    matches.push(doc)
  }

  const doc = matches.join('\n')
  return doc.length ? doc : null;
}

export function loadAndMergeQueryDocuments(inputPaths: string[], tagName: string = 'gql'): DocumentNode {
  const sources = inputPaths.map(inputPath => {
    const body = fs.readFileSync(inputPath, 'utf8');
    if (!body) {
      return null;
    }

    if (inputPath.endsWith('.jsx') || inputPath.endsWith('.js')
      || inputPath.endsWith('.tsx') || inputPath.endsWith('.ts')
    ) {
      const doc = extractDocumentFromJavascript(body.toString(), tagName);
      return doc ? new Source(doc, inputPath) : null;
    }

    return new Source(body, inputPath);
  }).filter(source => source);

  return concatAST((sources as Source[]).map(source => parse(source)));
}
