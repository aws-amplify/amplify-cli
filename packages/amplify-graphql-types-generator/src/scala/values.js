import {
  join,
  wrap,
} from '../utilities/printing';

export function escapedString(string) {
  return string.replace(/"/g, '\\"');
}

export function multilineString(context, string) {
  const lines = string.split('\n');
  lines.forEach((line, index) => {
    const isLastLine = index != lines.length - 1;
    context.printOnNewline(`"${escapedString(line)}"` + (isLastLine ? ' +' : ''));
  });
}

export function dictionaryLiteralForFieldArguments(args) {
  function expressionFromValue(value) {
    if (value.kind === 'Variable') {
      return `Variable("${value.variableName}")`;
    } else if (Array.isArray(value)) {
      return wrap('[', join(value.map(expressionFromValue), ', '), ']');
    } else if (typeof value === 'object') {
      return wrap('[', join(Object.entries(value).map(([key, value]) => {
        return `"${key}": ${expressionFromValue(value)}`;
      }), ', ') || ':', ']');
    } else {
      return JSON.stringify(value);
    }
  }

  return wrap('[', join(args.map(arg => {
    return `"${arg.name}": ${expressionFromValue(arg.value)}`;
  }), ', ') || ':', ']');
}
