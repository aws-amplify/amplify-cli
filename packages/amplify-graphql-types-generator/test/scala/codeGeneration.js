import { stripIndent } from 'common-tags';

import {
  parse,
  isType,
  GraphQLID,
  GraphQLString,
  GraphQLInt,
  GraphQLList,
  GraphQLNonNull,
  GraphQLInputObjectType,
  GraphQLEnumType,
} from 'graphql';

import {
  classDeclarationForOperation,
  caseClassDeclarationForFragment,
  caseClassDeclarationForSelectionSet,
  typeDeclarationForGraphQLType,
} from '../../src/scala/codeGeneration';

import { dictionaryLiteralForFieldArguments } from '../../src/scala/values';

import { loadSchema } from '../../src/loading';
const schema = loadSchema(require.resolve('../fixtures/starwars/schema.json'));

import CodeGenerator from '../../src/utilities/CodeGenerator';

import { compileToLegacyIR } from '../../src/compiler/legacyIR';

describe('Scala code generation', function() {
  let generator;
  let resetGenerator;
  let compileFromSource;
  let addFragment;

  beforeEach(function() {
    resetGenerator = () => {
      const context = {
        schema: schema,
        operations: {},
        fragments: {},
        typesUsed: {},
      };
      generator = new CodeGenerator(context);
    };

    compileFromSource = (source, options = { generateOperationIds: false }) => {
      const document = parse(source);
      let context = compileToLegacyIR(schema, document);
      options.generateOperationIds && Object.assign(context, { generateOperationIds: true, operationIdsMap: {} });
      generator.context = context;
      return context;
    };

    addFragment = fragment => {
      generator.context.fragments[fragment.fragmentName] = fragment;
    };

    resetGenerator();
  });

  describe('#classDeclarationForOperation()', function() {
    test(`should generate a class declaration for a query with variables`, function() {
      const { operations, fragments } = compileFromSource(`
        query HeroName($episode: Episode) {
          hero(episode: $episode) {
            name
          }
        }
      `);

      classDeclarationForOperation(generator, operations['HeroName'], Object.values(fragments));
      expect(generator.output).toMatchSnapshot();
    });

    test(`should generate a class declaration for a query with fragment spreads`, function() {
      const { operations, fragments } = compileFromSource(`
        query Hero {
          hero {
            ...HeroDetails
          }
        }

        fragment HeroDetails on Character {
          name
        }
      `);

      classDeclarationForOperation(generator, operations['Hero'], Object.values(fragments));
      expect(generator.output).toMatchSnapshot();
    });

    test(`should generate a class declaration for a query with conditional fragment spreads`, function() {
      const { operations, fragments } = compileFromSource(`
        query Hero {
          hero {
            ...DroidDetails
          }
        }

        fragment DroidDetails on Droid {
          primaryFunction
        }
      `);

      classDeclarationForOperation(generator, operations['Hero'], Object.values(fragments));
      expect(generator.output).toMatchSnapshot();
    });

    test(`should generate a class declaration for a query with a fragment spread nested in an inline fragment`, function() {
      const { operations, fragments } = compileFromSource(`
        query Hero {
          hero {
            ... on Droid {
              ...HeroDetails
            }
          }
        }

        fragment HeroDetails on Character {
          name
        }
      `);

      classDeclarationForOperation(generator, operations['Hero'], Object.values(fragments));

      expect(generator.output).toMatchSnapshot();
    });

    test(`should generate a class declaration for a mutation with variables`, function() {
      const { operations, fragments } = compileFromSource(`
        mutation CreateReview($episode: Episode) {
          createReview(episode: $episode, review: { stars: 5, commentary: "Wow!" }) {
            stars
            commentary
          }
        }
      `);

      classDeclarationForOperation(generator, operations['CreateReview'], Object.values(fragments));

      expect(generator.output).toMatchSnapshot();
    });

    describe(`when generateOperationIds is specified`, function() {
      let compileOptions = { generateOperationIds: true };

      test(`should generate a class declaration with an operationId property`, function() {
        const context = compileFromSource(
          `
          query Hero {
            hero {
              ...HeroDetails
            }
          }
          fragment HeroDetails on Character {
            name
          }
        `,
          compileOptions
        );

        classDeclarationForOperation(generator, context.operations['Hero'], Object.values(context.fragments));
        expect(generator.output).toMatchSnapshot();
      });

      test(`should generate different operation ids for different operations`, function() {
        const context1 = compileFromSource(
          `
          query Hero {
            hero {
              ...HeroDetails
            }
          }
          fragment HeroDetails on Character {
            name
          }
        `,
          compileOptions
        );

        classDeclarationForOperation(generator, context1.operations['Hero'], Object.values(context1.fragments));
        const output1 = generator.output;

        resetGenerator();
        const context2 = compileFromSource(
          `
          query Hero {
            hero {
              ...HeroDetails
            }
          }
          fragment HeroDetails on Character {
            appearsIn
          }
        `,
          compileOptions
        );

        classDeclarationForOperation(generator, context2.operations['Hero'], Object.values(context2.fragments));
        const output2 = generator.output;

        expect(output1).not.toBe(output2);
      });

      test(`should generate the same operation id regardless of operation formatting/commenting`, function() {
        const context1 = compileFromSource(
          `
          query HeroName($episode: Episode) {
            hero(episode: $episode) {
              name
            }
          }
        `,
          compileOptions
        );

        classDeclarationForOperation(generator, context1.operations['HeroName'], Object.values(context1.fragments));
        const output1 = generator.output;

        resetGenerator();
        const context2 = compileFromSource(
          `
          # Profound comment
          query HeroName($episode:Episode) { hero(episode: $episode) { name } }
          # Deeply meaningful comment
        `,
          compileOptions
        );

        classDeclarationForOperation(generator, context2.operations['HeroName'], Object.values(context2.fragments));
        const output2 = generator.output;

        expect(output1).toBe(output2);
      });

      test(`should generate the same operation id regardless of fragment order`, function() {
        const context1 = compileFromSource(
          `
          query Hero {
            hero {
              ...HeroName
              ...HeroAppearsIn
            }
          }
          fragment HeroName on Character {
            name
          }
          fragment HeroAppearsIn on Character {
            appearsIn
          }
        `,
          compileOptions
        );

        classDeclarationForOperation(generator, context1.operations['Hero'], Object.values(context1.fragments));
        const output1 = generator.output;

        resetGenerator();
        const context2 = compileFromSource(
          `
          query Hero {
            hero {
              ...HeroName
              ...HeroAppearsIn
            }
          }
          fragment HeroAppearsIn on Character {
            appearsIn
          }
          fragment HeroName on Character {
            name
          }
        `,
          compileOptions
        );

        classDeclarationForOperation(generator, context2.operations['Hero'], Object.values(context2.fragments));
        const output2 = generator.output;

        expect(output1).toBe(output2);
      });

      test(`should generate appropriate operation id mapping source when there are nested fragment references`, function() {
        const source = `
          query Hero {
            hero {
              ...HeroDetails
            }
          }
          fragment HeroName on Character {
            name
          }
          fragment HeroDetails on Character {
            ...HeroName
            appearsIn
          }
        `;
        const context = compileFromSource(source, true);
        expect(context.operations['Hero'].sourceWithFragments).toMatchSnapshot();
      });
    });
  });

  describe('#caseClassDeclarationForFragment()', function() {
    test(`should generate a caseClass declaration for a fragment with an abstract type condition`, function() {
      const { fragments } = compileFromSource(`
        fragment HeroDetails on Character {
          name
          appearsIn
        }
      `);

      caseClassDeclarationForFragment(generator, fragments['HeroDetails']);

      expect(generator.output).toMatchSnapshot();
    });

    test(`should generate a caseClass declaration for a fragment with a concrete type condition`, function() {
      const { fragments } = compileFromSource(`
        fragment DroidDetails on Droid {
          name
          primaryFunction
        }
      `);

      caseClassDeclarationForFragment(generator, fragments['DroidDetails']);

      expect(generator.output).toMatchSnapshot();
    });

    test(`should generate a caseClass declaration for a fragment with a subselection`, function() {
      const { fragments } = compileFromSource(`
        fragment HeroDetails on Character {
          name
          friends {
            name
          }
        }
      `);

      caseClassDeclarationForFragment(generator, fragments['HeroDetails']);

      expect(generator.output).toMatchSnapshot();
    });

    test(`should generate a caseClass declaration for a fragment that includes a fragment spread`, function() {
      const { fragments } = compileFromSource(`
        fragment HeroDetails on Character {
          name
          ...MoreHeroDetails
        }

        fragment MoreHeroDetails on Character {
          appearsIn
        }
      `);

      caseClassDeclarationForFragment(generator, fragments['HeroDetails']);

      expect(generator.output).toMatchSnapshot();
    });
  });

  describe('#caseClassDeclarationForSelectionSet()', function() {
    test(`should generate a caseClass declaration for a selection set`, function() {
      caseClassDeclarationForSelectionSet(generator, {
        caseClassName: 'Hero',
        parentType: schema.getType('Character'),
        fields: [
          {
            responseName: 'name',
            fieldName: 'name',
            type: GraphQLString,
          },
        ],
      });

      expect(generator.output).toMatchSnapshot();
    });

    test(`should escape reserved keywords in a caseClass declaration for a selection set`, function() {
      caseClassDeclarationForSelectionSet(generator, {
        caseClassName: 'Hero',
        parentType: schema.getType('Character'),
        fields: [
          {
            responseName: 'private',
            fieldName: 'name',
            type: GraphQLString,
          },
        ],
      });

      expect(generator.output).toMatchSnapshot();
    });

    test(`should generate a nested caseClass declaration for a selection set with subselections`, function() {
      caseClassDeclarationForSelectionSet(generator, {
        caseClassName: 'Hero',
        parentType: schema.getType('Character'),
        fields: [
          {
            responseName: 'friends',
            fieldName: 'friends',
            type: new GraphQLList(schema.getType('Character')),
            fields: [
              {
                responseName: 'name',
                fieldName: 'name',
                type: GraphQLString,
              },
            ],
          },
        ],
      });

      expect(generator.output).toMatchSnapshot();
    });

    test(`should generate a caseClass declaration for a selection set with a fragment spread that matches the parent type`, function() {
      addFragment({
        fragmentName: 'HeroDetails',
        typeCondition: schema.getType('Character'),
      });

      caseClassDeclarationForSelectionSet(generator, {
        caseClassName: 'Hero',
        parentType: schema.getType('Character'),
        fragmentSpreads: ['HeroDetails'],
        fields: [
          {
            responseName: 'name',
            fieldName: 'name',
            type: GraphQLString,
          },
        ],
      });

      expect(generator.output).toMatchSnapshot();
    });

    test(`should generate a caseClass declaration for a selection set with a fragment spread with a more specific type condition`, function() {
      addFragment({
        fragmentName: 'DroidDetails',
        typeCondition: schema.getType('Droid'),
      });

      caseClassDeclarationForSelectionSet(generator, {
        caseClassName: 'Hero',
        parentType: schema.getType('Character'),
        fragmentSpreads: ['DroidDetails'],
        fields: [
          {
            responseName: 'name',
            fieldName: 'name',
            type: GraphQLString,
          },
        ],
      });

      expect(generator.output).toMatchSnapshot();
    });

    test(`should generate a caseClass declaration for a selection set with an inline fragment`, function() {
      caseClassDeclarationForSelectionSet(generator, {
        caseClassName: 'Hero',
        parentType: schema.getType('Character'),
        fields: [
          {
            responseName: 'name',
            fieldName: 'name',
            type: new GraphQLNonNull(GraphQLString),
          },
        ],
        inlineFragments: [
          {
            typeCondition: schema.getType('Droid'),
            possibleTypes: ['Droid'],
            fields: [
              {
                responseName: 'name',
                fieldName: 'name',
                type: new GraphQLNonNull(GraphQLString),
              },
              {
                responseName: 'primaryFunction',
                fieldName: 'primaryFunction',
                type: GraphQLString,
              },
            ],
          },
        ],
      });

      expect(generator.output).toMatchSnapshot();
    });

    test(`should generate a caseClass declaration for a fragment spread nested in an inline fragment`, function() {
      addFragment({
        fragmentName: 'HeroDetails',
        typeCondition: schema.getType('Character'),
      });

      caseClassDeclarationForSelectionSet(generator, {
        caseClassName: 'Hero',
        parentType: schema.getType('Character'),
        fields: [],
        inlineFragments: [
          {
            typeCondition: schema.getType('Droid'),
            possibleTypes: ['Droid'],
            fields: [],
            fragmentSpreads: ['HeroDetails'],
          },
        ],
      });

      expect(generator.output).toMatchSnapshot();
    });
  });

  describe('#dictionaryLiteralForFieldArguments()', function() {
    test('should include expressions for input objects with variables', function() {
      const { operations } = compileFromSource(`
        mutation FieldArgumentsWithInputObjects($commentary: String!, $red: Int!) {
          createReview(episode: JEDI, review: { stars: 2, commentary: $commentary, favorite_color: { red: $red, blue: 100, green: 50 } }) {
            commentary
          }
        }
      `);

      const fieldArguments = operations['FieldArgumentsWithInputObjects'].fields[0].args;
      const dictionaryLiteral = dictionaryLiteralForFieldArguments(fieldArguments);

      expect(dictionaryLiteral).toBe(
        '["episode": "JEDI", "review": ["stars": 2, "commentary": Variable("commentary"), "favorite_color": ["red": Variable("red"), "blue": 100, "green": 50]]]'
      );
    });
  });

  describe('#typeDeclarationForGraphQLType()', function() {
    test('should generate an enum declaration for a GraphQLEnumType', function() {
      const generator = new CodeGenerator();

      typeDeclarationForGraphQLType(generator, schema.getType('Episode'));

      expect(generator.output).toMatchSnapshot();
    });

    test('should escape identifiers in cases of enum declaration for a GraphQLEnumType', function() {
      const generator = new CodeGenerator();

      const albumPrivaciesEnum = new GraphQLEnumType({
        name: 'AlbumPrivacies',
        values: { PUBLIC: { value: 'PUBLIC' }, PRIVATE: { value: 'PRIVATE' } },
      });

      typeDeclarationForGraphQLType(generator, albumPrivaciesEnum);

      expect(generator.output).toMatchSnapshot();
    });

    test('should generate a caseClass declaration for a GraphQLInputObjectType', function() {
      const generator = new CodeGenerator();

      typeDeclarationForGraphQLType(generator, schema.getType('ReviewInput'));

      expect(generator.output).toMatchSnapshot();
    });
  });
});
