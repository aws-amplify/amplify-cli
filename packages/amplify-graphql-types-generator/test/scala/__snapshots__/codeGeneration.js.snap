// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scala code generation #caseClassDeclarationForFragment() should generate a caseClass declaration for a fragment that includes a fragment spread 1`] = `
"case class HeroDetails(name: String, appearsIn: Seq[Option[String]]) extends me.shadaj.slinky.core.WithRaw {
}

object HeroDetails {
  val possibleTypes = scala.collection.Set(\\"Human\\", \\"Droid\\")
  implicit def toMoreHeroDetails(a: HeroDetails): MoreHeroDetails = MoreHeroDetails(a.appearsIn)
  val fragmentString =
    \\"fragment HeroDetails on Character {\\" +
    \\"  name\\" +
    \\"  ...MoreHeroDetails\\" +
    \\"}\\"
}"
`;

exports[`Scala code generation #caseClassDeclarationForFragment() should generate a caseClass declaration for a fragment with a concrete type condition 1`] = `
"case class DroidDetails(name: String, primaryFunction: Option[String]) {
}

object DroidDetails {
  val possibleTypes = scala.collection.Set(\\"Droid\\")
  val fragmentString =
    \\"fragment DroidDetails on Droid {\\" +
    \\"  name\\" +
    \\"  primaryFunction\\" +
    \\"}\\"
}"
`;

exports[`Scala code generation #caseClassDeclarationForFragment() should generate a caseClass declaration for a fragment with a subselection 1`] = `
"case class HeroDetails(name: String, friends: Option[Seq[Option[Friend]]]) extends me.shadaj.slinky.core.WithRaw {
}

object HeroDetails {
  val possibleTypes = scala.collection.Set(\\"Human\\", \\"Droid\\")
  val fragmentString =
    \\"fragment HeroDetails on Character {\\" +
    \\"  name\\" +
    \\"  friends {\\" +
    \\"    name\\" +
    \\"  }\\" +
    \\"}\\"
}


case class Friend(name: String) extends me.shadaj.slinky.core.WithRaw {
}

object Friend {
  val possibleTypes = scala.collection.Set(\\"Human\\", \\"Droid\\")
}"
`;

exports[`Scala code generation #caseClassDeclarationForFragment() should generate a caseClass declaration for a fragment with an abstract type condition 1`] = `
"case class HeroDetails(name: String, appearsIn: Seq[Option[String]]) extends me.shadaj.slinky.core.WithRaw {
}

object HeroDetails {
  val possibleTypes = scala.collection.Set(\\"Human\\", \\"Droid\\")
  val fragmentString =
    \\"fragment HeroDetails on Character {\\" +
    \\"  name\\" +
    \\"  appearsIn\\" +
    \\"}\\"
}"
`;

exports[`Scala code generation #caseClassDeclarationForSelectionSet() should escape reserved keywords in a caseClass declaration for a selection set 1`] = `
"case class Hero(private: Option[String]) extends me.shadaj.slinky.core.WithRaw {
}

object Hero {
  val possibleTypes = scala.collection.Set(\\"Human\\", \\"Droid\\")
}"
`;

exports[`Scala code generation #caseClassDeclarationForSelectionSet() should generate a caseClass declaration for a fragment spread nested in an inline fragment 1`] = `
"case class Hero() extends me.shadaj.slinky.core.WithRaw {
  def asDroid: Option[AsDroid] = {
    if (AsDroid.possibleTypes.contains(this.raw.__typename.asInstanceOf[String])) Some(implicitly[me.shadaj.slinky.core.Reader[AsDroid]].read(this.raw)) else None
  }
}

case class AsDroid() {
}

object AsDroid {
  val possibleTypes = scala.collection.Set(\\"Droid\\")
  implicit def toHero(a: AsDroid): Hero = Hero()
  implicit def toHeroDetails(a: AsDroid): HeroDetails = HeroDetails()
}

object Hero {
  val possibleTypes = scala.collection.Set(\\"Human\\", \\"Droid\\")
}"
`;

exports[`Scala code generation #caseClassDeclarationForSelectionSet() should generate a caseClass declaration for a selection set 1`] = `
"case class Hero(name: Option[String]) extends me.shadaj.slinky.core.WithRaw {
}

object Hero {
  val possibleTypes = scala.collection.Set(\\"Human\\", \\"Droid\\")
}"
`;

exports[`Scala code generation #caseClassDeclarationForSelectionSet() should generate a caseClass declaration for a selection set with a fragment spread that matches the parent type 1`] = `
"case class Hero(name: Option[String]) extends me.shadaj.slinky.core.WithRaw {
}

object Hero {
  val possibleTypes = scala.collection.Set(\\"Human\\", \\"Droid\\")
  implicit def toHeroDetails(a: Hero): HeroDetails = HeroDetails()
}"
`;

exports[`Scala code generation #caseClassDeclarationForSelectionSet() should generate a caseClass declaration for a selection set with a fragment spread with a more specific type condition 1`] = `
"case class Hero(name: Option[String]) extends me.shadaj.slinky.core.WithRaw {
  def asDroidDetails: Option[DroidDetails] = {
    if (DroidDetails.possibleTypes.contains(this.raw.__typename.asInstanceOf[String])) Some(implicitly[me.shadaj.slinky.core.Reader[DroidDetails]].read(this.raw)) else None
  }
}

object Hero {
  val possibleTypes = scala.collection.Set(\\"Human\\", \\"Droid\\")
}"
`;

exports[`Scala code generation #caseClassDeclarationForSelectionSet() should generate a caseClass declaration for a selection set with an inline fragment 1`] = `
"case class Hero(name: String) extends me.shadaj.slinky.core.WithRaw {
  def asDroid: Option[AsDroid] = {
    if (AsDroid.possibleTypes.contains(this.raw.__typename.asInstanceOf[String])) Some(implicitly[me.shadaj.slinky.core.Reader[AsDroid]].read(this.raw)) else None
  }
}

case class AsDroid(name: String, primaryFunction: Option[String]) {
}

object AsDroid {
  val possibleTypes = scala.collection.Set(\\"Droid\\")
  implicit def toHero(a: AsDroid): Hero = Hero(a.name)
}

object Hero {
  val possibleTypes = scala.collection.Set(\\"Human\\", \\"Droid\\")
}"
`;

exports[`Scala code generation #caseClassDeclarationForSelectionSet() should generate a nested caseClass declaration for a selection set with subselections 1`] = `
"case class Hero(friends: Option[Seq[Option[Friend]]]) extends me.shadaj.slinky.core.WithRaw {
}

object Hero {
  val possibleTypes = scala.collection.Set(\\"Human\\", \\"Droid\\")
}


case class Friend(name: Option[String]) extends me.shadaj.slinky.core.WithRaw {
}

object Friend {
  val possibleTypes = scala.collection.Set(\\"Human\\", \\"Droid\\")
}"
`;

exports[`Scala code generation #classDeclarationForOperation() should generate a class declaration for a mutation with variables 1`] = `
"object CreateReviewMutation extends com.apollographql.scalajs.GraphQLMutation {
  val operationString =
    \\"mutation CreateReview($episode: Episode) {\\" +
    \\"  createReview(episode: $episode, review: {stars: 5, commentary: \\\\\\"Wow!\\\\\\"}) {\\" +
    \\"    stars\\" +
    \\"    commentary\\" +
    \\"  }\\" +
    \\"}\\"
  val operation = com.apollographql.scalajs.gql(operationString)


  case class Variables(episode: Option[String]) {
  }

  case class Data(createReview: Option[CreateReview]) {
  }

  object Data {
    val possibleTypes = scala.collection.Set(\\"Mutation\\")
  }

  case class CreateReview(stars: Int, commentary: Option[String]) {
  }

  object CreateReview {
    val possibleTypes = scala.collection.Set(\\"Review\\")
  }
}"
`;

exports[`Scala code generation #classDeclarationForOperation() should generate a class declaration for a query with a fragment spread nested in an inline fragment 1`] = `
"object HeroQuery extends com.apollographql.scalajs.GraphQLQuery {
  val operationString =
    \\"query Hero {\\" +
    \\"  hero {\\" +
    \\"    ... on Droid {\\" +
    \\"      ...HeroDetails\\" +
    \\"    }\\" +
    \\"  }\\" +
    \\"}\\"

  val requestString: String = { operationString + HeroDetails.fragmentString }
  val operation = com.apollographql.scalajs.gql(requestString)

  type Variables = Unit

  case class Data(hero: Option[Hero]) {
  }

  object Data {
    val possibleTypes = scala.collection.Set(\\"Query\\")
  }


  case class Hero() extends me.shadaj.slinky.core.WithRaw {
    def asDroid: Option[AsDroid] = {
      if (AsDroid.possibleTypes.contains(this.raw.__typename.asInstanceOf[String])) Some(implicitly[me.shadaj.slinky.core.Reader[AsDroid]].read(this.raw)) else None
    }
  }

  case class AsDroid(name: String) {
  }

  object AsDroid {
    val possibleTypes = scala.collection.Set(\\"Droid\\")
    implicit def toHero(a: AsDroid): Hero = Hero()
    implicit def toHeroDetails(a: AsDroid): HeroDetails = HeroDetails(a.name)
  }

  object Hero {
    val possibleTypes = scala.collection.Set(\\"Human\\", \\"Droid\\")
  }
}"
`;

exports[`Scala code generation #classDeclarationForOperation() should generate a class declaration for a query with conditional fragment spreads 1`] = `
"object HeroQuery extends com.apollographql.scalajs.GraphQLQuery {
  val operationString =
    \\"query Hero {\\" +
    \\"  hero {\\" +
    \\"    ...DroidDetails\\" +
    \\"  }\\" +
    \\"}\\"

  val requestString: String = { operationString + DroidDetails.fragmentString }
  val operation = com.apollographql.scalajs.gql(requestString)

  type Variables = Unit

  case class Data(hero: Option[Hero]) {
  }

  object Data {
    val possibleTypes = scala.collection.Set(\\"Query\\")
  }


  case class Hero() extends me.shadaj.slinky.core.WithRaw {
    def asDroid: Option[AsDroid] = {
      if (AsDroid.possibleTypes.contains(this.raw.__typename.asInstanceOf[String])) Some(implicitly[me.shadaj.slinky.core.Reader[AsDroid]].read(this.raw)) else None
    }
    def asDroidDetails: Option[DroidDetails] = {
      if (DroidDetails.possibleTypes.contains(this.raw.__typename.asInstanceOf[String])) Some(implicitly[me.shadaj.slinky.core.Reader[DroidDetails]].read(this.raw)) else None
    }
  }

  case class AsDroid(primaryFunction: Option[String]) {
  }

  object AsDroid {
    val possibleTypes = scala.collection.Set(\\"Droid\\")
    implicit def toHero(a: AsDroid): Hero = Hero()
    implicit def toDroidDetails(a: AsDroid): DroidDetails = DroidDetails(a.primaryFunction)
  }

  object Hero {
    val possibleTypes = scala.collection.Set(\\"Human\\", \\"Droid\\")
  }
}"
`;

exports[`Scala code generation #classDeclarationForOperation() should generate a class declaration for a query with fragment spreads 1`] = `
"object HeroQuery extends com.apollographql.scalajs.GraphQLQuery {
  val operationString =
    \\"query Hero {\\" +
    \\"  hero {\\" +
    \\"    ...HeroDetails\\" +
    \\"  }\\" +
    \\"}\\"

  val requestString: String = { operationString + HeroDetails.fragmentString }
  val operation = com.apollographql.scalajs.gql(requestString)

  type Variables = Unit

  case class Data(hero: Option[Hero]) {
  }

  object Data {
    val possibleTypes = scala.collection.Set(\\"Query\\")
  }


  case class Hero(name: String) extends me.shadaj.slinky.core.WithRaw {
  }

  object Hero {
    val possibleTypes = scala.collection.Set(\\"Human\\", \\"Droid\\")
    implicit def toHeroDetails(a: Hero): HeroDetails = HeroDetails(a.name)
  }
}"
`;

exports[`Scala code generation #classDeclarationForOperation() should generate a class declaration for a query with variables 1`] = `
"object HeroNameQuery extends com.apollographql.scalajs.GraphQLQuery {
  val operationString =
    \\"query HeroName($episode: Episode) {\\" +
    \\"  hero(episode: $episode) {\\" +
    \\"    name\\" +
    \\"  }\\" +
    \\"}\\"
  val operation = com.apollographql.scalajs.gql(operationString)


  case class Variables(episode: Option[String]) {
  }

  case class Data(hero: Option[Hero]) {
  }

  object Data {
    val possibleTypes = scala.collection.Set(\\"Query\\")
  }


  case class Hero(name: String) extends me.shadaj.slinky.core.WithRaw {
  }

  object Hero {
    val possibleTypes = scala.collection.Set(\\"Human\\", \\"Droid\\")
  }
}"
`;

exports[`Scala code generation #classDeclarationForOperation() when generateOperationIds is specified should generate a class declaration with an operationId property 1`] = `
"object HeroQuery extends com.apollographql.scalajs.GraphQLQuery {
  val operationString =
    \\"query Hero {\\" +
    \\"  hero {\\" +
    \\"    ...HeroDetails\\" +
    \\"  }\\" +
    \\"}\\"

  val operationIdentifier: String = \\"90d0d674eb6a7b33776f63200d6cec3d09f881247c360a2ac9a29037a02210c4\\"

  val requestString: String = { operationString + HeroDetails.fragmentString }
  val operation = com.apollographql.scalajs.gql(requestString)

  type Variables = Unit

  case class Data(hero: Option[Hero]) {
  }

  object Data {
    val possibleTypes = scala.collection.Set(\\"Query\\")
  }


  case class Hero(name: String) extends me.shadaj.slinky.core.WithRaw {
  }

  object Hero {
    val possibleTypes = scala.collection.Set(\\"Human\\", \\"Droid\\")
    implicit def toHeroDetails(a: Hero): HeroDetails = HeroDetails(a.name)
  }
}"
`;

exports[`Scala code generation #classDeclarationForOperation() when generateOperationIds is specified should generate appropriate operation id mapping source when there are nested fragment references 1`] = `
"query Hero {
  hero {
    ...HeroDetails
  }
}
fragment HeroDetails on Character {
  ...HeroName
  appearsIn
}
fragment HeroName on Character {
  name
}"
`;

exports[`Scala code generation #typeDeclarationForGraphQLType() should escape identifiers in cases of enum declaration for a GraphQLEnumType 1`] = `
"object AlbumPrivacies {
  val public = \\"PUBLIC\\"
  val private = \\"PRIVATE\\"
}
"
`;

exports[`Scala code generation #typeDeclarationForGraphQLType() should generate a caseClass declaration for a GraphQLInputObjectType 1`] = `
"/**
 * The input object sent when someone is creating a new review
 */
case class ReviewInput(stars: Int, commentary: Option[String], favoriteColor: Option[ColorInput]) {
}"
`;

exports[`Scala code generation #typeDeclarationForGraphQLType() should generate an enum declaration for a GraphQLEnumType 1`] = `
"/**
 * The episodes in the Star Wars trilogy
 */
object Episode {
  /**
   * Star Wars Episode IV: A New Hope, released in 1977.
   */
  val newhope = \\"NEWHOPE\\"
  /**
   * Star Wars Episode V: The Empire Strikes Back, released in 1980.
   */
  val empire = \\"EMPIRE\\"
  /**
   * Star Wars Episode VI: Return of the Jedi, released in 1983.
   */
  val jedi = \\"JEDI\\"
}
"
`;
